### DISCUSSION TITLE
How can the SRI help KPs?

### DISCUSSION THREAD

**Discussion by cbizon** [NCATSTranslator/TranslatorArchitecture#32] [General]
Suppose that KPs will implement some simple deductive reasoning.  What SRI tooling could help?  Library? API? what kind of signature? What are the requirements?

Here are a few things that already exist, which could be used as is, modified, repackaged or other.

* The [Biolink Model Toolkit](https://github.com/biolink/biolink-model-toolkit) is a python library that can be used to simply interrogate the biolink model.  So you can for instance find out all the superclasses of ChemicalSubstance or all the subproperties of affects.

* The [Biolink Lookup Service](https://bl-lookup-sri.renci.org/apidocs/) wraps the above toolkit in a simple web service.  If it is going to be heavily used, it would need some work to scale a bit better than it does at the moment.

* [This library](https://github.com/ranking-agent/reasoner) from the ranking agent converts TRAPI queries to cypher, incorporating the predicate reasoning.

Do those spur any ideas?  KPs, are any of these helpful?  What would make them more useful?

**Comment by cbizon** [comment]
And one more:

* The [Ontology KP](https://stars-app.renci.org/sparql-kp) uses OBO ontologies to determine superclass_of and subclass_of for entities.

**Comment by ehinderer** [comment]
I don't have any experience using any of these but they all sound promising as tools to leverage for simple predicate inferencing. 

As discussed in #34 I think all that would be needed in terms of inferencing for the time being would be lookups for ontological descendants and reflexive predicates. 

Along these lines, something that came up in one of our internal meetings was the issue that TRAPI message validation doesn't check for for BioLink compliant contents, including predicates (at least I think it still doesn't). So there'd be a risk that if KPs used the Biolink toolkit for this kind of thing there'd be name resolution issues potentially. Someone correct me if I'm wrong.

**Reply by cbizon** [reply]
No, you're correct - there is some biolink compliance built into KGX, but I'm thinking maybe we need to either make a service for that or merge it with the trapi validator ...
